src/exchange/providers/retry_with_back_off_decorator.py

```diff
@@ -0,0 +1,51 @@
+from functools import wraps
+import time
+from typing import Any, Callable, Dict, List, Optional
+from httpx import HTTPStatusError, Response
+
+def retry_with_backoff(
+        should_retry: Callable, 
+        max_retries: Optional[int] = 5, 
+        initial_wait: Optional[float] = 10, 
+        backoff_factor: Optional[float] = 1, 
+        handle_retry_exhausted: Optional[Callable] = None) ->  Callable:
+    def decorator(func: Callable) -> Callable:
+        @wraps(func)
+        def wrapper(*args: List, **kwargs: Dict) -> Any:  # noqa: ANN401
+            result = None
+            for retry in range(max_retries):
+                result = func(*args, **kwargs)
+                if not should_retry(result):
+                    return result
+                if (retry + 1) == max_retries:
+                    break
+                sleep_time = initial_wait + (backoff_factor * (2 ** retry))
+                time.sleep(sleep_time)
+            handle_retry_exhausted(result, max_retries)
```

Comment:
handle_retry_exhausted is Optional, so this will fail if the default (None) is used